### ISSUE TITLE
Adding unnecessary `use libp2p;` statement leads to unrelated error when adding String to &str

### ISSUE THREAD

**Issue by Frederik-Baetens** [rust-lang/rust#92209]
**I tried this code:**

```rust
use libp2p;
fn main() {
    let teststring = "hi".to_string();
    println!("{}", String::from("/") + &teststring);
}
```
Repo with code sample: https://github.com/Frederik-Baetens/libp2p-err-sample

The used crate does need to be libp2p. I haven't found any other crates with which this behavior occurs. I understand that this points to weird stuff going on in libp2p, but I think that strange behavior like this should never happen, no matter what black magic is going on in libp2p.

**I expected to see this happen:** No compiler error when running `cargo check`

**Instead, this happened:**
```
error[E0277]: cannot add `&String` to `String`
 --> src/main.rs:4:41
  |
4 |     let addedstring = String::from("/") + &teststring;
  |                                         ^ no implementation for `String + &String`
  |
  = help: the trait `Add<&String>` is not implemented for `String`
```

The error only occurs when the uneccesary use statement is present.

### Meta

`rustc --version --verbose`:
```
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0
```
I also tried with nightly and the bug exists there as well.

This bug also happened in the gitlab CI for the project I originally encountered this bug in.

**Comment by Patrick-Poitras** [comment]
Seems that this bug is triggered by multiple libraries. See #92178, which is near identical except for the library used.

**Comment by compiler-errors** [comment]
I started investigating this in the other issue #92178.

**Comment by compiler-errors** [comment]
haha, okay so this is due to the fact that there's currently only one stdlib candidate `Add<_> for String` in scope. This means when we collect relevant impls for Add, there's only one until you add that `use` (or define an impl locally like I did in the repro minimization).

Adding another relevant impl in scope will cause us to skip the short-circuit in [`SelectionContext::candidate_from_obligation_no_cache` @ L153](https://github.com/rust-lang/rust/blob/master/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs#L153), and so we actually try winnowing the (>1) candidates, which fails to account for autoref coersion in the parameter.. leading to zero impls and the failure here.

**Comment by Frederik-Baetens** [comment]
Closing as duplicate then.
